tket/src/WeightSubgrMono/Searching/SearchBranch.cpp

```diff
@@ -0,0 +1,309 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/Searching/SearchBranch.hpp"
+
+#include <algorithm>
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/GeneralUtils.hpp"
+#include "WeightSubgrMono/EndToEndWrappers/SolutionData.hpp"
+#include "WeightSubgrMono/GraphTheoretic/NeighboursData.hpp"
+#include "WeightSubgrMono/Reducing/DistancesReducer.hpp"
+#include "WeightSubgrMono/WeightPruning/WeightChecker.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+SearchBranch::SearchBranch(
+    const DomainInitialiser::InitialDomains& initial_domains,
+    const NeighboursData& pattern_ndata, NearNeighboursData& pattern_near_ndata,
+    const NeighboursData& target_ndata, NearNeighboursData& target_near_ndata,
+    unsigned max_distance_reduction_value, ExtraStatistics& extra_statistics)
+    : m_pattern_ndata(pattern_ndata),
+      m_target_ndata(target_ndata),
+      m_extra_statistics(extra_statistics),
+      m_derived_graphs_reducer(m_pattern_ndata, m_target_ndata),
+      m_neighbours_reducer(m_pattern_ndata, m_target_ndata),
+      m_nodes_raw_data_wrapper(initial_domains),
+      m_domains_accessor(m_nodes_raw_data_wrapper),
+      m_node_list_traversal(m_nodes_raw_data_wrapper) {
+  m_extra_statistics.number_of_pattern_vertices =
+      m_pattern_ndata.get_number_of_nonisolated_vertices();
+  m_extra_statistics.number_of_target_vertices =
+      m_target_ndata.get_number_of_nonisolated_vertices();
+  m_extra_statistics.initial_number_of_possible_assignments = 0;
+  for (const auto& entry : initial_domains) {
+    m_extra_statistics.initial_number_of_possible_assignments += entry.size();
+  }
+
+  // In what order should we do reduction/checks?
+  // The simplest/cheapest first? Most powerful?
+  // Seems a difficult question...
+  m_reducer_wrappers.emplace_back(m_neighbours_reducer);
```

Comment:
I was thinking that we'd have many different kinds of reducers. OK, right now there are just 3 types (distance reducers also). I wanted a single unified interface. Unfortunately the Hall set reduction doesn't quite fit in, but all the others do.

However, there are more actual reducer objects (a different one for each distance). A bit crude, but the simplest way - we DON'T want to check a single domain over ALL distances d=1,2,3,...,N. Instead, for each distance d=1,2,3,... we check ALL domains. A subtle difference in order, but I believe should be faster - you want to detect nogoods ASAP to back out early, and smaller distances first should find incompatibilities more quickly.

I don't see what you're asking. Reducer wrappers are just ways to refer to reducer objects. HOWEVER they cannot contain them directly as (1) the reducers are different types (2) the reducers have different constructors. So, even if the wrappers used unique ptrs to contain reducer objects, I don't think it would be as clean as it is currently.